docs/tutorial/codecs.txt

```diff
@@ -0,0 +1,433 @@
+encode/decode it:
+
+.. code-block:: php
+
+   <?php
+
+   final class Person
+   {
+       public MongoDB\BSON\ObjectId $id;
+       public string $name;
+
+       public function __construct(string $name)
+       {
+           $this->id = new MongoDB\BSON\ObjectId();
+           $this->name = $name;
+       }
+   }
+
+.. code-block:: php
+
+   <?php
+
+   final class PersonCodec implements MongoDB\Codec\DocumentCodec
+   {
+       // These traits define commonly used functionality to avoid duplication
+       use MongoDB\Codec\DecodeIfSupported;
+       use MongoDB\Codec\EncodeIfSupported;
+
+       public function canDecode($value): bool
+       {
+           return $value instanceof MongoDB\BSON\Document && $value->has('name');
+       }
+
+       public function canEncode($value): bool
+       {
+           return $value instanceof Person;
+       }
+
+       public function decode($value): Person
+       {
+           if (! $this->canDecode($value)) {
+               throw UnsupportedValueException::invalidDecodableValue($value);
+           }
+
+           $person = new Person($value->get('name'));
+           $person->id = $value->get('_id');
+
+           return $person;
+       }
+
+       public function encode($value): MongoDB\BSON\Document
+       {
+           if (! $this->canEncode($value)) {
+               throw UnsupportedValueException::invalidEncodableValue($value);
+           }
+
+           return MongoDB\BSON\Document::fromPHP([
+               '_id' => $value->id,
+               'name' => $value->name,
+           ]);
+       }
+   }
+
+To then use this codec with a collection, specify the ``codec`` option when selecting the collection:
+
+.. code-block:: php
+
+   <?php
+
+   $client = new MongoDB\Client();
+   $collection = $client->selectCollection('test', 'person', [
+       'codec' => new PersonCodec(),
+   ]);
+
+   $person = new Person('Jane Doe');
+   $collection->insertOne($person);
+
+   $person = $collection->findOne();
+
+The example above selects a collection and instructs it to use the ``PersonCodec`` for encoding and decoding documents.
+When inserting data, the ``PersonCodec`` is used to encode the document. When retrieving data, the ``PersonCodec`` is
+used to decode BSON data into a ``Person`` instance. Note that while the ``PersonCodec`` could technically decode any
+BSON document that contains a name field, we wouldn't want to use it for any other documents. Document codecs are meant
+to be used in a collection, or when decoding embedded documents.
+
+When working on a collection with a codec, the codec will only accept and return data of that type for certain
+operations. The ``bulkWrite``, ```findOneAndReplace``, ``insertMany``, ``insertOne``, and ``replaceOne`` operations
+will attempt to encode the given data using the provided codec. Trying to insert or replace a document that cannot be
+encoded will result in an exception. The ``aggregate``, ``find``, ``findOne``, ``findOneAndDelete``,
```

Comment:
Didn't we discuss _not_ having ``aggregate`` apply the collection's codec under the assumption that a document may be modified by the pipeline? I just checked the source and `Collection::aggregate()` definitely inherits the codec with `inheritCodecOrTypeMap()`.

Perhaps that discussion was only about change streams, `distinct()`, etc.

Reply:
I remember discussing `aggregate` in that context, but don't remember a decision on it. I'm on the fence about it and can see arguments made in both directions, so I'm not too sure what the best course of action is.

Reply:
That's easier to remove the codec by setting `'codec' => null` than injecting the Codec instance to where the line `Collection::aggregate is called`. There is plenty of use-cases for using aggregate to return documents as in the collection (`$search`, `$geoNear`).

Reply:
I believe we also talked this through over video. SGTM. I'll resolve.